src/transaction.js

```diff
@@ -196,61 +300,38 @@ Transaction.fromBuffer = function (buffer, network = networks.bitcoin, __noStric
     tx.expiryHeight = readUInt32()
   }
 
+  if (coins.isZcash(network) && tx.version >= Transaction.ZCASH_SAPLING_VERSION) {
+    tx.valueBalance = readUInt64()
+  }
+
   if (coins.isZcash(network) && tx.version >= Transaction.ZCASH_JOINSPLITS_SUPPORT_VERSION) {
-    var joinSplitsLen = readVarInt()
-    for (i = 0; i < joinSplitsLen; ++i) {
-      var vpubOld = readUInt64()
-      var vpubNew = readUInt64()
-      var anchor = readSlice(32)
-      var nullifiers = []
-      for (var j = 0; j < Transaction.ZCASH_NUM_JOINSPLITS_INPUTS; j++) {
-        nullifiers.push(readSlice(32))
-      }
-      var commitments = []
-      for (j = 0; j < Transaction.ZCASH_NUM_JOINSPLITS_OUTPUTS; j++) {
-        commitments.push(readSlice(32))
-      }
-      var ephemeralKey = readSlice(32)
-      var randomSeed = readSlice(32)
-      var macs = []
-      for (j = 0; j < Transaction.ZCASH_NUM_JOINSPLITS_INPUTS; j++) {
-        macs.push(readSlice(32))
-      }
-      // TODO what are those exactly? Can it be expressed by BigNum?
-      var zproof = {
-        gA: readCompressedG1(),
-        gAPrime: readCompressedG1(),
-        gB: readCompressedG2(),
-        gBPrime: readCompressedG1(),
-        gC: readCompressedG1(),
-        gCPrime: readCompressedG1(),
-        gK: readCompressedG1(),
-        gH: readCompressedG1()
-      }
-      var ciphertexts = []
-      for (j = 0; j < Transaction.ZCASH_NUM_JOINSPLITS_OUTPUTS; j++) {
-        ciphertexts.push(readSlice(Transaction.ZCASH_NOTECIPHERTEXT_SIZE))
+    if (tx.version >= Transaction.ZCASH_SAPLING_VERSION) {
+      var nShieldedSpend = readVarInt()
+      for (i = 0; i < nShieldedSpend; ++i) {
+        tx.vShieldedSpend.push(readShieldedSpend())
       }
 
-      tx.joinsplits.push({
-        vpubOld: vpubOld,
-        vpubNew: vpubNew,
-        anchor: anchor,
-        nullifiers: nullifiers,
-        commitments: commitments,
-        ephemeralKey: ephemeralKey,
-        randomSeed: randomSeed,
-        macs: macs,
-        zproof: zproof,
-        ciphertexts: ciphertexts
-      })
+      var nShieldedOutput = readVarInt()
+      for (i = 0; i < nShieldedOutput; ++i) {
+        tx.vShieldedOutput.push(readShieldedOutput())
+      }
+    }
+    var joinSplitsLen = readVarInt()
```

Comment:
This is where I think the second half of that if statement should be put. I think this would make the code cleaner as it would be clear what version each part of the raw tx is from.